File: 3725.count-ways-to-choose-coprime-integers-from-rows.py
#
# @lc app=leetcode id=3725 lang=python3
#
# [3725] Count Ways to Choose Coprime Integers from Rows
#

from typing import List
from math import gcd
from functools import lru_cache

class Solution:
    def countCoprime(self, mat: List[List[int]]) -> int:
        MOD = 10**9 + 7
        m, n = len(mat), len(mat[0])
        
        @lru_cache(None)
        def dfs(row: int, current_gcd: int) -> int:
            if row == m:  # If we've processed all rows
                return 1 if current_gcd == 1 else 0
            
            count = 0
            # Try each number in the current row
            for num in mat[row]:
                new_gcd = gcd(current_gcd, num)
                # Only continue if there's still a chance to reach gcd of 1
                if new_gcd <= current_gcd:
                    count += dfs(row + 1, new_gcd)
                    count %= MOD
            return count
        
        total_ways = dfs(0, 0)
        return total_ways % MOD

File: 3725.count-ways-to-choose-coprime-integers-from-rows.go
#
# @lc app=leetcode id=3725 lang=python3
#
# [3725] Count Ways to Choose Coprime Integers from Rows
#

# @lc code=start
from math import gcd
from collections import defaultdict

MOD = int(1e9 + 7)

class Solution:
    def countCoprime(self, mat):
        m = len(mat)
        n = len(mat[0])
        
        # Initialize DP table
        dp = [defaultdict(int) for _ in range(m + 1)]
        dp[0][0] = 1 # base case: choosing nothing has gcd of zero (neutral element)
        
        # Iterate over each row in matrix
        for i in range(m):
            for num in mat[i]:
                for g, count in list(dp[i].items()):
                    new_gcd = gcd(g, num)
                    dp[i + 1][new_gcd] = (dp[i + 1][new_gcd] + count) % MOD
        
        return dp[m][1]
# @lc code=end